File: year2/semester2/mpp/c#/TripClientServer/TripClientServer/Service/TripServicesProxy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TripCommon.domain;
using TripCommon.Dto;
using TripCommon.Net;
using TripCommon.Validator;

namespace TripClient.Service
{
    class TripServicesProxy : BaseServicesProxy, ITripServices
    {
        protected ITripObserver tripObserver;
        public TripServicesProxy(string host, int port) : base(host, port)
        {
        }

        public void AddBooking(string client, string phone, int numTickets, Trip trip, Account account)
        {
            LOGGER.InfoFormat("adding booking for client {0}  fo trip {1} tickets {2}", client, trip, numTickets);
            EnsureConnected();
            TripDto t = new TripDto(trip.Id, trip.Landmark, trip.CompanyName, trip.DepartureTime, trip.Price, trip.AvailablePlaces);
            AccountDto a = new AccountDto(account.Id, account.Name, account.Password);
            Request r = new Request()
            {
                Type = RequestType.ADD_BOOKING,
                Data = new BookingDto(client, phone, numTickets, t, a)
            };
            SendRequest(r);
            Response response = ReadResponse();
            LOGGER.InfoFormat("response for add booking {0}", response);
            if (response.Type == ResponseType.OK)
            {
                LOGGER.Info("booking added");
                return;
            }
            if (response.Type == ResponseType.ERROR)
            {
                String err = response.Data.ToString();
                LOGGER.Info("received ERROR response " + err);
                throw new ServiceException(err);
            }
        }



        public Account FindAccountByNameAndPassword(string name, string password)
        {
            LOGGER.InfoFormat("finding account with name {0} password {1}", name, password);
            EnsureConnected();
            Request r = new Request()
            {
                Type = RequestType.LOGIN,
                Data = new AccountDto(name, password)
            };
            SendRequest(r);
            Response response = ReadResponse();
            LOGGER.InfoFormat("response for find account received is {0}", response);
            if (response.Type == ResponseType.OK)
            {
                AccountDto accountDto = (AccountDto)response.Data;
                if (accountDto == null)
                {
                    LOGGER.Info("No account found");
                    return null;
                }
                Account account = new Account(accountDto.id, accountDto.name, accountDto.password);
                LOGGER.InfoFormat("found account {0}", account);
                return account;
            }
            if (response.Type == ResponseType.ERROR)
            {
                String err = response.Data.ToString();
                LOGGER.Info("finding account failed received ERROR response " + err);
                throw new ServiceException(err);
            }
            return null;
        }


        public List<Trip> GetAllTrips()
        {
            LOGGER.InfoFormat("getting all trips");
            EnsureConnected();
            Request r = new Request()
            {
                Type = RequestType.GET_ALL_TRIPS
            };
            SendRequest(r);
            Response response = ReadResponse();
            LOGGER.InfoFormat("response for get all trips received is {0}", response);
            if (response.Type == ResponseType.OK)
            {
                List<TripDto> tripDtos = (List<TripDto>)response.Data;
                List<Trip> trips = tripDtos
                    .Select(t => new Trip(t.id, t.landmark, t.companyName, t.departure, t.price, t.places))
                    .ToList();
                LOGGER.Info("found all trips");
                return trips;
            }
            if (response.Type == ResponseType.ERROR)
            {
                String err = response.Data.ToString();
                LOGGER.Info("getting all trips failed ERROR response " + err);
                throw new ServiceException(err);
            }
            return null;
        }

        public List<Trip> GetTripsByLandmarkDepartureHour(string landmark, int start, int end)
        {
            LOGGER.InfoFormat("getting trips by landmark {0} departure hours {1} {2}", landmark, start, end);
            EnsureConnected();
            Request r = new Request()
            {
                Type = RequestType.SEARCH_TRIPS,
                Data = new TripDto(landmark, start, end)
            };
            SendRequest(r);
            Response response = ReadResponse();
            LOGGER.InfoFormat("response for search trips received is {0}", response);
            if (response.Type == ResponseType.OK)
            {
                List<TripDto> tripDtos = (List<TripDto>)response.Data;
                List<Trip> trips = tripDtos
                    .Select(t => new Trip(t.id, t.landmark, t.companyName, t.departure, t.price, t.places))
                    .ToList();
                LOGGER.Info("found all trips");
                return trips;
            }
            if (response.Type == ResponseType.ERROR)
            {
                String err = response.Data.ToString();
                LOGGER.Info("searching trips failed ERROR response " + err);
                throw new ServiceException(err);
            }
            return null;
        }

        public void Logout()
        {
            LOGGER.Info("logging out");
            if (connection == null)
            {
                LOGGER.Info("connection already closed");
                return;
            }
            Request r = new Request()
            {
                Type = RequestType.LOGOUT
            };
            SendRequest(r);
            Response response = ReadResponse();
            CloseConnection();
            if (response.Type == ResponseType.ERROR)
            {
                String err = response.Data.ToString();
                LOGGER.Warn("logging out failed " + err);
                throw new ServiceException(err);
            }
        }
        public void AddTripObserver(ITripObserver observer)
        {
            tripObserver = observer;
        }
        public void RemoveTripObserver(ITripObserver observer)
        {
            tripObserver = null;
        }

        public override void HandleBookingAdded(Response response)
        {
            LOGGER.InfoFormat("handling booking added response {0}", response);
            BookingDto bookingDto = (BookingDto)response.Data;
            Booking booking = new Booking(bookingDto.id, bookingDto.trip.id, bookingDto.account.id, bookingDto.client
                , bookingDto.phone, bookingDto.numTickets);
            tripObserver.BookingInserted(booking);
        }
    }
}

File: year2/semester2/mpp/java/TripClientServer/TripClient/src/main/java/service/TripServicesProxy.java
package service;

import domain.Account;
import domain.Booking;
import domain.Trip;
import dto.AccountDto;
import dto.BookingDto;
import dto.TripDto;
import net.Request;
import net.RequestType;
import net.Response;
import net.ResponseType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class TripServicesProxy extends BaseServiceProxy implements TripServices {
    private static final Logger LOGGER = LogManager.getLogger(TripServicesProxy.class.getName());
    protected TripObserver tripObserver;

    public TripServicesProxy(String host, int port) {
        super(host, port);
    }

    @Override
    public Account getAccountByNamePassword(String name, String password) {
        LOGGER.traceEntry("getting account by name " + name + " and password " + password);
        ensureConnected();
        Request req = new Request.Builder()
                .type(RequestType.LOGIN)
                .data(new AccountDto(name, password))
                .build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.OK) {
            LOGGER.info("received OK response");
            AccountDto accountDto = (AccountDto) response.data();
            if (accountDto == null) {
                LOGGER.info("no account found");
                return null;
            }
            Account a = new Account(accountDto.id, accountDto.name, accountDto.password);
            LOGGER.info("found account " + a);
            return a;
        }
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            LOGGER.info("received ERROR response " + err);
            throw new ServiceException(err);
        }
        LOGGER.traceExit();
        return null;
    }

    @Override
    public void addBooking(String client, String phone, int numTickets, Trip trip, Account account) {
        LOGGER.traceEntry("adding booking for client " + client + " to trip " + trip + " made by " + account);
        ensureConnected();
        TripDto tripDto = new TripDto(trip.getId(), trip.getLandmark(), trip.getCompanyName(), trip.getDepartureTime(),
                                      trip.getPrice(), trip.getAvailablePlaces());
        AccountDto accountDto = new AccountDto(account.getId(), account.getName(), account.getPassword());
        Request req = new Request.Builder()
                .type(RequestType.ADD_BOOKING)
                .data(new BookingDto(client, phone, numTickets, tripDto, accountDto))
                .build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.OK) {
            LOGGER.info("received OK response");
            LOGGER.info("booking added");
            return;
        }
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            LOGGER.info("add booking failed received ERROR response " + err);
//            closeConnection();
            throw new ServiceException(err);
        }
        LOGGER.traceExit();
    }

    @Override
    public List<Trip> getAllTrips() {
        LOGGER.info("getting all trips");
        ensureConnected();
        Request req = new Request.Builder()
                .type(RequestType.GET_ALL_TRIPS)
                .build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.OK) {
            LOGGER.info("getting all trips OK response");
            List<TripDto> trips = (ArrayList<TripDto>) response.data();
            return trips.stream()
                    .map(t -> new Trip(t.id, t.landmark, t.companyName, t.departureTime, t.price, t.availablePlaces))
                    .collect(Collectors.toList());
        }
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            LOGGER.warn("getting all trips failed");
//            closeConnection();
            throw new ServiceException(err);
        }
        return null;
    }

    @Override
    public List<Trip> getTripsByLandmarkDepartureHour(String landmark, int start, int end) {
        LOGGER.traceEntry("getting trips with landmark " + landmark + " and between " + start + " " + end);
        ensureConnected();
        ArrayList<Integer> list = new ArrayList<Integer>();
        list.add(start);
        list.add(end);
        Request req = new Request.Builder()
                .type(RequestType.SEARCH_TRIPS)
                .data(new TripDto(landmark, list))
                .build();
        sendRequest(req);
        Response response = readResponse();
        LOGGER.info("response for search trips " + response);
        if (response.type() == ResponseType.OK) {
            LOGGER.info("received OK response");
            List<TripDto> trips = (ArrayList<TripDto>) response.data();
            LOGGER.info("returning searched trips ");
            return trips.stream()
                    .map(t -> new Trip(t.id, t.landmark, t.companyName, t.departureTime, t.price, t.availablePlaces))
                    .collect(Collectors.toList());
        }
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            LOGGER.info("searching trips failed received ERROR response " + err);
            throw new ServiceException(err);
        }
        LOGGER.traceExit();
        return null;
    }

    @Override
    public void logOut() {
        LOGGER.info("logging out");
//        ensureConnected();
        if (connection == null){
            LOGGER.info("connection already closed");
            return;
        }
        Request req = new Request.Builder().type(RequestType.LOGOUT).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            LOGGER.warn("logging out failed " + err);
            throw new ServiceException(err);
        }
    }

    protected void handleBookingAdded(Response response) {
        BookingDto bookingDto = (BookingDto)response.data();
        Booking b = new Booking(bookingDto.id, bookingDto.account.id, bookingDto.trip.id, bookingDto.client, bookingDto.phone, bookingDto.numTickets);
        tripObserver.bookingInserted(b);
    }
    @Override
    public void addTripObserver(TripObserver observer) {
        tripObserver = observer;
    }

    @Override
    public void removeTripObserver(TripObserver observer) {
        tripObserver = null;
    }
}
